enhancements/coreos-encrypted-by-default.md

```diff
@@ -0,0 +1,422 @@
+
+Ignition is the first boot agent for FCOS and RHCOS. FCOS discussed the topic of partition layouts, with the idea of having Ignition lay-down arbitrary root filesystem layouts. At this time, Openshift requires Ignition V2, while FCOS uses Ignition v3; these versions are incompatible. When Ignition supports the required depenendencies, a reviewion to this proposal will be forthcoming.
+
+## Terms
+* LUKS/LUKS2: the standard Linux encryption method. Encrypted filesystems stored in a container.
+* Master Key: the _actual_ encryption key used. The master key encrypted using a passphrase and is embedded in the LUKS meta-data on disk encrypted. This encryption key is unknown to users.
+* LUKS. Passphrase: any string or file that can be used to unlock the master key. Each passphrase is bound to a "key-slot."
+* LUKS Token: arbitrary JSON stored in the LUKS meta-data.
+* NBE: [Network Bound-Disk Encryption](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Security_Guide/sec-Using_Network-Bound_Disk_Encryption.html). A method of providing context-based disk encryption.
+* [Clevis](https://github.com/latchset/clevis): Linux CLI tools for policy-based encryption and decryption. Clevis has a Dracut module for automatically opening LUKS containers in an NBE context.
+* Clevis pin: a program called by Clevis to encrypt and decryption of secrets.
+* Tang: A server implementation that provides cryptographic binding services without the need for an escrow. Secrets are derived from what the client and the Tang server both have, reducing the attack surface.
+* KMS: Key Management System, such as Amazon KMS.
+* [TPM2](https://en.wikipedia.org/wiki/Trusted_Platform_Module): A secure cryptoprocessor for storing keys in hardware.
+* Dracut Module: program(s) that run early in boot before the on-disk Linux is booted.
+
+If you are unfamiliar with some of these terms [please see an excellent Youtube video on Clevis/Tang](https://youtu.be/2uLKvB8Z5D0).
+
+## Proposal
+
+To provide install-time _and_ rolling re-encryption, CoreOS boot images will:
+* Have the root filesystem housed in a LUKS container.
+* Use the 'null-cipher' for the initial LUKS container "encrpytion". This allows _any_ passphrase to unlocked the LUKS container.
+* The passphrase will be stored in a LUKS token via the Clevis plaintext pin.
+
+Note: the default boot image is _effectively *NOT* encrypted_.
+
+On first boot, a Dracut module will re-encrypt the LUKS container unless the kernel CLI parameter ('rd.coreos.no_encrpytion=1') is used.  The module will likely run after Ignition has completed. It will:
+  * Prepare a random passphrase,
+  * Randomize the LUKS container's UUID,
+  * Re-encrypt the LUKS container using AES-256-CBC (FIPS 140-2 compliant). This process generates a new random master key.
+  * If a Clevis configration is found, apply it in the initramfs. Otherwise:
```

Comment:
Where would the clevis configuration come from?

Bigger picture...I think this enhancement needs to start from e.g. AWS IPI via `openshift-install` - how does the admin configure it there?

Reply:
Either Ignition or the MCO would be the _channel_ for delivering the Clevis configuration. Part of the design is that the initial Clevis configuration is not the final configuration. To provide the install time configuration, the user would run `openshift-install create ignition-configs` and update the assets to include the Clevis configuration.

In the early stages of discussing this, @crawford pointed out that the installer should not know everything possible OS-level option. Having the Installer generate the Clevis configuration would be extremely difficult (KSM, Tang, TPM?) and would be quite situational.

Reply:
> To provide the install time configuration, the user would run openshift-install create ignition-configs and update the assets to include the Clevis configuration.

That doesn't work for IPI though.  I think a model where the config is represented as a MachineConfig object might work better.

This is what I was thinking of for [fcct](https://github.com/coreos/fcct):

```
storage:
  root-tpm: true
``` or so to bind the rootfs to a TPM, (and require one exist).

We could have something similar in the `MachineConfig` object.

(However, anything of this form gets tricky; we'd probably need the MCS to transcode to Ignition so that it applies in the initramfs via Ignition or so)

Reply:
The installer will need to support it for initial configuration. The default Clevis for TPM2 or if we use the plain-text pin will be generated by the Dracut module.

Reply:
> The installer will need to support it for initial configuration. 

Specifically, are you proposing adding a section to the [install-config.yaml](https://github.com/openshift/installer/blob/d9a9648cf2330d467cca9f2988846d031464125e/docs/user/customization.md#platform-customization) ?